### PR TITLE
Feature/validation aspnet

### PR DESCRIPTION
**Pull request by humulla**
## Link to JIRA ticket

[Jira Ticket](https://hackney.atlassian.net/browse/TL-153)
## Describe this PR

Created a new package called Hackney.Core.Validation.AspNet and added all relevant code 

#### _Checklist_

- [x] Added tests to cover all new production code
- [x] Added comments to the README or updated relevant documentation _(add link to documentation)_, where necessary.
- [x] Checked all code for possible refactoring
- [x] Code pipeline builds correctly

### CODE REVIEW

File: Hackney.Core/Hackney.Core.Validation.AspNet/HttpRequestExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+        /// <returns>The request's body text</returns>
+        public static async Task<string> GetRawBodyStringAsync(this HttpRequest request, Encoding encoding = null)
+        {
+            if (encoding == null)
+                encoding = Encoding.UTF8;
+
+            request.Body.Seek(0, SeekOrigin.Begin);
+            using (StreamReader reader = new StreamReader(request.Body, encoding))
+                return await reader.ReadToEndAsync().ConfigureAwait(false);
+        }
```
Comment (mikemoody-amido) [inline Hackney.Core/Hackney.Core.Validation.AspNet/HttpRequestExtensions.cs]: I think that this should go in the Hackney.Core.Http package and that package should be pulled in here.

File: Hackney.Core/Hackney.Core.Validation.AspNet/Hackney.Core.Validation.AspNet.csproj [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+  <PropertyGroup>
+	<TargetFramework>netcoreapp3.1</TargetFramework>
+  </PropertyGroup>
+  
+  <PropertyGroup>
+    <RepositoryUrl>https://github.com/LBHackney-IT/lbh-core</RepositoryUrl>
+    <Authors>London Borough of Hackney Council</Authors>
+    <Company>London Borough of Hackney Council</Company>
+    <Copyright>2021</Copyright>
+    <Description>Common library providing property validation rules to use in LBH Web API services.</Description>
```
Comment (mikemoody-amido) [inline Hackney.Core/Hackney.Core.Validation.AspNet/Hackney.Core.Validation.AspNet.csproj]: Not just rules, but help configuring the use of FluentValidation.

File: Hackney.Core/Hackney.Core.Http/HttpRequestExtensions.cs [hunk 3]
```diff
@@ -3,7 +3,7 @@
 using System.Text;
 using System.Threading.Tasks;
 
-namespace Hackney.Core.Validation.AspNet
+namespace Hackney.Core.Http
 {
     public static class HttpRequestExtensions
     {
```
Comment (mikemoody-amido) [inline Hackney.Core/Hackney.Core.Http/HttpRequestExtensions.cs]: That's not what I meant.
This class should not be in this project at all, but in the Hackney.Core.Http project.

File: Hackney.Core/Hackney.Core.Validation.AspNet/ReadMe.md [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+Fluent Validation allows you to separate validation rules from your object model and helps you structure the rules so that they are nice and readable. The rules are also super easy to test.
+
+#### UseErroCodeInterceptor
+
+We have customised our FluentValidation through the use of an interceptor. The `UseErrorCodeInterceptor` implements the `IValidatorInterceptor`.
+
+This interface has two methods – `BeforeAspNetValidation` and `AfterAspNetValidation`.
+
+`BeforeMvcValidation` is invoked after the appropriate validator has been selected but before it is invoked. One of the arguments passed to this method is a `ValidationContext` that will eventually be passed to the validator. The context has several properties including a reference to the object being validated. If we want to change which rules are going to be invoked (for example, by using a custom `ValidatorSelector`) then we can create a new `ValidationContext`, set its Selector property, and return that from the `BeforeAspNetValidation` method.
```
Comment (mikemoody-amido) [inline Hackney.Core/Hackney.Core.Validation.AspNet/ReadMe.md]: I wouldn't focus on how this class works internally (as it is called by the MVC framework and not directly by any client code), but what it is used for and examples of how it is used.

File: Hackney.Core/Hackney.Core.Validation.AspNet/ReadMe.md [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+
+For example
+
+```csharp
+
+   services.AddFluentValidation(fv => fv.RegisterValidatorsFromAssemblies(assemblies));
+   services.TryAddTransient<IValidatorInterceptor, UseErrorCodeInterceptor>();
+  
+```
+The example above customises all validators to use the `AddFluentValidation()` as well as the `IValidatorInteceptor`
```
Comment (mikemoody-amido) [inline Hackney.Core/Hackney.Core.Validation.AspNet/ReadMe.md]: Not really. Yes it does register the UseErroCodeInterceptor but more than that it registers all the validators located in either the current assembly, or the specified assemblies.
Provide examples of how to call it using both overloads

File: README.md [hunk 6]
```diff
@@ -44,6 +44,7 @@ The following features are implemented within this package.
   * [Shared Classes](#Shared-Classes)
 * [Validation](#Validation)
   * [XssValidator](#XssValidator)
+* [Validation.AspNet](#Hackney.Core/Hackney.Core.Validation.AspNet/ReadMe.md)
```
Comment (mikemoody-amido) [inline README.md]: Should start with a / and not  # (I think)

File: Hackney.Core/Hackney.Core.Http/HttpRequestExtensions.cs (hunk 3 above)
Comment (humulla) [reply Hackney.Core/Hackney.Core.Http/HttpRequestExtensions.cs]: This has been moved to the Http Project

File: Hackney.Core/Hackney.Core.Validation.AspNet/ReadMe.md [hunk 7]
```diff
@@ -6,24 +6,54 @@ Fluent Validation allows you to separate validation rules from your object model
 
 #### UseErroCodeInterceptor
```
Comment (mikemoody-amido) [inline Hackney.Core/Hackney.Core.Validation.AspNet/ReadMe.md]: Typo

File: Hackney.Core/Hackney.Core.Validation.AspNet/ReadMe.md [hunk 8]
```diff
@@ -6,24 +6,54 @@ Fluent Validation allows you to separate validation rules from your object model
...
+         {
+             var projection = result.Errors.Select(failure => ConstructFailure(failure));
+             return new ValidationResult(projection);
+         }
+
+         return result;
+      }
+```
+The first method `ConstructFailure` constructs the error message that would be provided if there is a validation error. 
+The second method `AfterAspNetValidation` displays the error message when there is a validation error.
```
Comment (mikemoody-amido) [inline Hackney.Core/Hackney.Core.Validation.AspNet/ReadMe.md]: Still not quite right (and I may be partly to blame here as I didn't explain fully the first time).
The `UseErrorCodeInterceptor` class is used under the covers to make sure that any error code that is specified on a validation rule is actually included in the HTTP response. By default it just includes the error message and any code specified is left out. Clients should not need to invoke this class directly.

So when a validation rule does this...
```
RuleFor(x => x.ReasonForTermination).NotXssString()
             .WithErrorCode(ErrorCodes.XssCheckFailure);
```
then the error code value for `XssCheckFailure` will actually be included in the error details in the HTTP response.

File: Hackney.Core/Hackney.Core.Validation.AspNet/ReadMe.md [hunk 9]
```diff
@@ -6,24 +6,54 @@ Fluent Validation allows you to separate validation rules from your object model
...
 
 ```csharp
 
+   public static IServiceCollection AddFluentValidation(this IServiceCollection services)
+   {
+      return services.AddFluentValidation(Assembly.GetExecutingAssembly());
+   }
+
    services.AddFluentValidation(fv => fv.RegisterValidatorsFromAssemblies(assemblies));
    services.TryAddTransient<IValidatorInterceptor, UseErrorCodeInterceptor>();
```
Comment (mikemoody-amido) [inline Hackney.Core/Hackney.Core.Validation.AspNet/ReadMe.md]: This example is not quite right...

```csharp
public void ConfigureServices(IServiceCollection services)
{
    ...

    // This example will register any validators just in the application's assembly
    services.AddFluentValidation();

    // This example will register any validators in the assembly where the CreatePersonRequestObjectValidator is located
    // (but not the local assembly - if you need that then add it specifically.)
    services.AddFluentValidation(Assembly.GetAssembly(typeof(CreatePersonRequestObjectValidator)));

    ...
}

```

File: Hackney.Core/Hackney.Core.Validation.AspNet/ReadMe.md [hunk 10]
```diff
@@ -47,13 +26,19 @@ Usage
...
-   services.TryAddTransient<IValidatorInterceptor, UseErrorCodeInterceptor>();
+    // This example will register any validators in the assembly where the CreatePersonRequestObjectValidator is located
+    // (but not the local assembly - if you need that then add it specifically.)
+    services.AddFluentValidation(Assembly.GetAssembly(typeof(CreatePersonRequestObjectValidator)));
+
+    ...
+   }
   
 ```
 The example above registers all the validators located in either the current assembly, or the specified assemblies
```
Comment (mikemoody-amido) [inline Hackney.Core/Hackney.Core.Validation.AspNet/ReadMe.md]: Probably not needed any more.

File: Hackney.Core/Hackney.Core.Validation.AspNet/ReadMe.md [hunk 11]
```diff
@@ -4,40 +4,19 @@ The main use of this packages is to configure the use of FluentValidation within
...
-      }
+      RuleFor(x => x.ReasonForTermination).NotXssString()
+                    .WithErrorCode(ErrorCodes.XssCheckFailure);
+
 ```
-The first method `ConstructFailure` constructs the error message that would be provided if there is a validation error. 
-The second method `AfterAspNetValidation` displays the error message when there is a validation error.
+then the error code value for XssCheckFailure will actually be included in the error details in the HTTP response.
 
 #### FluentValidationExtensions
```
Comment (mikemoody-amido) [inline Hackney.Core/Hackney.Core.Validation.AspNet/ReadMe.md:21]: Prob don't need this line at the top here.

**Review by mikemoody-amido — APPROVED** [review]

**Pull request merged**